lib/Analysis/ControlDependenceAnalysis.cpp

```diff
@@ -0,0 +1,298 @@
+      for (auto &one_dep_dep : one_dep_deps) {
+        // add this dep if it is not already present
+        if (std::find(block_deps.begin(), block_deps.end(), one_dep_dep) ==
+            block_deps.end())
+          control_deps_map[&block].push_back(one_dep_dep);
+      }
+    }
+  }
+}
+
+void ControlDependenceAnalysis::identifyForwardControlDeps(FuncOp &funcOp) {
+  Region &funcReg = funcOp.getRegion();
+  DominanceInfo domInfo;
+  llvm::DominatorTreeBase<Block, false> &domTree = domInfo.getDomTree(&funcReg);
+  // Get loop information to eliminate loop exits from the dependencies
+  CFGLoopInfo li(domTree);
+
+  DenseMap<Block *, SmallVector<Block *, 4>> control_deps_map;
+
+  // Initialize the control_deps_maps by creating an entry for every block
+  // constituting the region
+  for (Block &block : funcReg.getBlocks()) {
+    SmallVector<Block *, 4> deps;
+    control_deps_map.insert(std::make_pair(&block, deps));
+  }
+
+  for (Block &block : funcReg.getBlocks()) {
+    // Extract the dependencies of this block to adjust them by removing loop
+    // exit conditions, if any
+    // We adopt a convention of accessing the very last map of the vector
+    // because we only call this function at the end of
+    // identifyAllControlDeps() right after it adds a new entry to
+    // all_control_deps_maps
+    SmallVector<Block *, 4> block_deps =
+        all_control_deps_maps[all_control_deps_maps.size() - 1][&block];
+
+    for (auto &one_dep : block_deps) {
+      CFGLoop *loop = li.getLoopFor(one_dep);
+      if (loop == nullptr) {
+        // indicating that the one_dep is not inside any loop, so it must be a
+        // forward dependency
+        control_deps_map[&block].push_back(one_dep);
+      } else {
+        // indicating that the one_dep is inside a loop,
+        // to decide if it is a forward dep or not, compare it against all of
+        // the exits and latches of this loop
+        bool not_forward = false;
+        // check if one_dep is an exit of the loop
+        SmallVector<Block *> loop_exitBlocks;
+        loop->getExitingBlocks(loop_exitBlocks);
+        for (auto &loop_exit : loop_exitBlocks) {
+          if (loop_exit == one_dep) {
+            not_forward = true;
+            break; // it is not a forward dependency so no need to contiue
+                   // looping
+          }
+        }
+        // check if one_dep is a latch of the loop
+        SmallVector<Block *> loop_latchBlocks;
+        loop->getLoopLatches(loop_latchBlocks);
+        for (auto &loop_latch : loop_latchBlocks) {
+          if (loop_latch == one_dep) {
+            not_forward = true;
+            break; // it is not a forward dependency so no need to contiue
+                   // looping
+          }
+        }
+
+        if (!not_forward) {
+          control_deps_map[&block].push_back(one_dep);
+        }
+      }
+    }
+  }
+
+  forward_control_deps_maps.emplace_back(control_deps_map);
```

Comment:
Same comment as in `identifyAllControlDeps` about not copying the map. Use the reference returned by `emplace_back` inside the method to avoid a copy.